Bkd index

piotrrzysko opened the pull request:
## Summary of the changes / Why this improves CrateDB

Resolves to #11993.

While this work is not yet complete, I wanted to gather early feedback from you to ensure I'm heading in the right direction. In the commit messages, I've included links to the commits I backported from Elasticsearch.

### Benchmarks
I've' created a benchmark that compares prefix and BKD indexes, which you can find here: https://github.com/piotrrzysko/crate-benchmarks/commit/987616f716772fd4e85aba2bb9f97670e4636b86. Below, you'll find the results from running this benchmark on my laptop:

```
## Running Query:
   Statement:
     insert into shaped (id, shape) values (?, ?)
   Concurrency: 1
   Iterations: 1000
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [01:12<00:00, 13.83 requests/s]
Runtime (in ms):
    mean:    69.531 ± 1.080
    min/max: 38.073 → 166.199
Percentile:
    50:   63.049 ± 17.424 (stdev)
    95:   103.257
    99.9: 165.550

## Running Query:
   Statement:
     select id from shaped where match(shape, ?)
   Concurrency: 1
   Iterations: 1000
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:12<00:00, 81.71 requests/s]
Runtime (in ms):
    mean:    10.662 ± 3.897
    min/max: 0.920 → 1927.001
Percentile:
    50:   3.230 ± 62.868 (stdev)
    95:   34.263
    99.9: 254.781

## Running Query:
   Statement:
     insert into bkd_shaped (id, shape) values (?, ?)
   Concurrency: 1
   Iterations: 1000
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [01:04<00:00, 15.53 requests/s]
Runtime (in ms):
    mean:    61.555 ± 0.701
    min/max: 37.428 → 131.462
Percentile:
    50:   57.692 ± 11.307 (stdev)
    95:   80.024
    99.9: 129.559

## Running Query:
   Statement:
     select id from bkd_shaped where match(shape, ?)
   Concurrency: 1
   Iterations: 1000
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:05<00:00, 186.21 requests/s]
Runtime (in ms):
    mean:    4.101 ± 2.997
    min/max: 0.567 → 1497.535
Percentile:
    50:   1.064 ± 48.350 (stdev)
    95:   6.529
    99.9: 247.819
```

### Remaining work
* Documentation update.
* Increase code coverage with tests. I think it would be good to cover more shapes and match types in GeoShapeIntegrationTest. 

### Questions

#### array(geo_shape INDEX USING ...)
I've noticed that arrays of geo_shapes with parameters are currently not supported (or I'm missing something). A statement like: 

```sql
CREATE TABLE shaped (shapes array(geo_shape INDEX using QUADTREE with (precision='1m', distance_error_pct='0.25')));
```

returns the following error: ```[XX000] ERROR: line 1:45: missing ')' at 'INDEX'```. Therefore, I cannot test BKD-tree indexes for arrays as the following statement:

```sql
CREATE TABLE shaped (shapes array(geo_shape INDEX using BKDTREE));
```

returns the same error. Should I create an issue for that or try to fix it in this PR?

#### GeoShapeFieldMapper
I wanted to test changes in GeoShapeFieldMapper. If I'm not mistaken this mapper is only used during index recovery. Therefore, I added a test in AlterTableRerouteIntegrationTest. However, I'm not sure if this is the best place. What do you think?  

## Checklist

 - [ ] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [ ] Updated documentation & `sql_features` table for user facing changes
 - [ ] Touched code is covered by tests
 - [X] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [X] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

crate-jenkins commented:
Can one of the admins verify this patch?

BaurzhanSakhariev commented:
ok to test

BaurzhanSakhariev commented on server/src/main/java/io/crate/metadata/GeoReference.java:
```diff
@@ -86,7 +81,12 @@ public GeoReference(ReferenceIdent ident,
+            throw new IllegalArgumentException(
+                "The parameters precision, tree_levels, and distance_error_pct are not applicable to BKD tree indexes."
+            );
+        }
```
I suggest to move this check to analyzer (we create `GeoReference` in `TableElementsAnalyzer.RefBuilder.build`). 

Then we can replace `DDLIntegrationTest.testSettingsNotApplicableToBkdTreeIndexes` by more lightweight tests cases in `AlterTableAddColumnAnalyzerTest` and `CreateAlterTableStatementAnalyzerTest`

BaurzhanSakhariev replied:
yep, let's keep the check in ctor but rewrite as  2 unit test - create table and add column.

piotrrzysko replied:
I replaced `DDLIntegrationTest.testSettingsNotApplicableToBkdTreeIndexes` with `GeoReferenceTest.testSettingsNotApplicableToBkdTreeIndexes`.

BaurzhanSakhariev commented on server/src/test/java/io/crate/integrationtests/AlterTableRerouteIntegrationTest.java:
```diff
@@ -56,4 +58,39 @@ public void testAlterTableRerouteMoveShard() throws Exception {
         assertThat(response).hasRowCount(0);
     }
+
+    @Test
+    public void testAlterTableRerouteMoveShardWithBkdTreeIndex() {
+        int shardId = 0;
+        String tableName = "my_table";
+        execute("create table " + tableName + " (" +
+            "id int primary key," +
+            "shape geo_shape INDEX USING bkdtree" +
+            ") clustered into 1 shards " +
+            "with (number_of_replicas=0)");
+        execute("INSERT INTO " + tableName + " (id, shape) VALUES (?, ?)", $$($(1L, "POINT (13.0 52.4)")));
+        execute("REFRESH TABLE " + tableName);
+        ensureGreen();
+        execute("select node['id'] from sys.shards where id = ? and table_name = ?", new Object[]{shardId, tableName});
+        String fromNode = (String) response.rows()[0][0];
+        execute("select id from sys.nodes where id != ?", new Object[]{fromNode});
+        String toNode = (String) response.rows()[0][0];
+
+        execute("ALTER TABLE my_table REROUTE MOVE SHARD ? FROM ? TO ?", new Object[]{shardId, fromNode, toNode});
+        assertThat(response).hasRowCount(1);
+        ensureGreen();
+        execute("select * from sys.shards where id = ? and node['id'] = ? and table_name = ?", new Object[]{shardId, toNode, tableName});
+        assertThat(response).hasRowCount(1);
+        execute("select * from sys.shards where id = ? and node['id'] = ? and table_name = ?", new Object[]{shardId, fromNode, tableName});
+        assertThat(response).hasRowCount(0);
+
+        execute("select id from " + tableName + " where match(shape, " +
+            "'POLYGON((12.998725175857544 52.40087142225922," +
+            "13.002265691757202 52.40087142225922," +
+            "13.002265691757202 52.39927416492016," +
+            "12.998725175857544 52.39927416492016," +
+            "12.998725175857544 52.40087142225922))')");
+        assertThat(response).hasRowCount(1);
+        assertThat(response).hasRows($(1));
+    }
```
FieldMappers are used for reading from the translog. 

This test can be replaced by a test, verifying that we can read geo shapes from the translog. Something similar to https://github.com/crate/crate/commit/2cc60d723932f710ead6525e2753512304fc2d55

mfussenegger replied:
We could extract parts of https://github.com/crate/crate/blob/ebfc70ae12bf1a26e88999930c3ba4235c66004c/server/src/testFixtures/java/io/crate/types/DataTypeTestCase.java#L76

into a re-usable functions where type definition/reference are parameterized.

piotrrzysko replied:
Thanks for the clarification. 

@mfussenegger would you mind taking a look at https://github.com/crate/crate/pull/14942/commits/2d7bab6074279f13968a60443ea03f87416454e5? I tried to extract some reusable components from the test you mentioned. Eventually, I moved the assertions that verify fields in both documents to a new Assert class and used the existing `getIndexer` method in both tests. Please let me know what you think about it.

BaurzhanSakhariev replied:
Looks good, thanks! 

Minor suggestion: we can make helper method in the `IndexerTest` public (slightly modify according to your change) instead of introducing a new class.

piotrrzysko replied:
Done. 

Previously, I introduced a separate class to avoid creating a circular dependency between the`test` and `testFixtures` directories, but perhaps it is not a problem.

BaurzhanSakhariev commented on server/src/main/java/io/crate/metadata/GeoReference.java:
```diff
@@ -48,7 +40,10 @@
+    public static final String GEOHASH_PREFIX_REE = "geohash";
+    public static final String LEGACY_QUAD_PREFIX_TREE = "legacyquadtree";
+    public static final String QUAD_PREFIX_TREE = "quadtree";
+    public static final String BKD_TREE = "bkdtree";
```
Minor: "geohash" and "quadtree" are already defined in `GeoShapeFieldMapper`, we could add new constants there to be consistent with the old code.

piotrrzysko replied:
Done.

BaurzhanSakhariev commented on server/src/test/java/io/crate/integrationtests/GeoShapeIntegrationTest.java:
```diff
@@ -63,14 +63,20 @@ public void prepare() throws Exception {
                 "  shape geo_shape," +
                 "  shapes array(geo_shape)" +
                 ") with (number_of_replicas=0)");
+        execute("CREATE TABLE bkd_shaped (" +
+                "  id long primary key," +
+                "  shape geo_shape INDEX USING bkdtree" +
+                ") with (number_of_replicas=0)");
```
Can we add a new field, indexed with bkdtree, into existing table instead of creating a new table?

BaurzhanSakhariev replied:
I'm now thinking whether it's a good idea to allow mixed indexing in a single table. Yes, as mentioned above, MatchPredicate  can be done BWC but not sure how build Lucene query in `WithinFunction.toQuery `for smth like `within(bkd_shape, geohash_shape)`. Even is possible, feels like error prone.

I think we should completely restrict non-BKD tree, mention that in breaking changes (that only BKD in supported from now on), add analyzer unit tests to verify that geohash/quadtree now throws an error. 

Then we don't need new table `bkd_shaped` and rather need to adjust all tests which are currently using geohash/quadtree to use BKD and they should be still passing.

However, we still need old indexing code in the `GeoShapeIndexer`/`GeoFieldMapper/GeoReference` for BWC reasons. We want to support rolling upgrades or insert into old tables, snapshotted in the old cluster and restored into new cluster). 

Maybe in 6.0 we can completely remove old code and say that we don't allow restore 5.* snapshots into 6.* clusters.

@seut, @mfussenegger, @matriv  wdyt?

BaurzhanSakhariev replied:
UPD:
`within(shape1, shape2)` falls to `GenericFunctionQuery` -> full scan anyway (see https://github.com/crate/crate/pull/14942#discussion_r1373106209)  but still believe we shouldn't mix old/new indexing.

BaurzhanSakhariev replied:
replied in https://github.com/crate/crate/pull/14942#issuecomment-1798656694

BaurzhanSakhariev commented on server/src/test/java/io/crate/integrationtests/GeoShapeIntegrationTest.java:
```diff
@@ -89,6 +95,16 @@ public void testSelectGeoShapeFromSource() throws Exception {
         assertThat(response.rows()[0][0], instanceOf(List.class));
     }
 
+    @Test
+    public void testSelectGeoShapeFromSource_bkdTree() {
```
Minor: let's use underscore for the whole test name (and in other similar changes).

BaurzhanSakhariev reviewed (commented):
Thanks a lot for your contribution!

Looks good in general. 

Regarding your questions:

1. 

> I've noticed that arrays of geo_shapes with parameters are currently not supported (or I'm missing something). 

It's supported, it was a syntax issue - `array(geo_shape)` must be used instead of `array(geo_shape ..... )`

While checking, I spot a bug, see https://github.com/crate/crate/pull/14956. Thanks for asking and helping to reveal the issue!

2. 
> I wanted to test changes in GeoShapeFieldMapper. If I'm not mistaken this mapper is only used during index recovery. Therefore, I added a test in AlterTableRerouteIntegrationTest.

FieldMappers are used to read from the translog. See comment below more more details and suggestion how to test it.

BaurzhanSakhariev commented on server/src/test/java/io/crate/integrationtests/DDLIntegrationTest.java:
```diff
@@ -259,6 +259,22 @@ public void testUnknownGeoShapeSetting() throws Exception {
             .hasHTTPError(BAD_REQUEST, 4000);
     }
 
+    @Test
+    public void testSettingsNotApplicableToBkdTreeIndexes() {
+        Asserts.assertSQLError(() -> execute("create table test (col1 geo_shape INDEX using BKDTREE with (precision='1m'))"))
+            .hasMessageContaining("The parameters precision, tree_levels, and distance_error_pct are not applicable to BKD tree indexes.")
+            .hasPGError(INTERNAL_ERROR)
+            .hasHTTPError(BAD_REQUEST, 4000);
+        Asserts.assertSQLError(() -> execute("create table test (col1 geo_shape INDEX using BKDTREE with (distance_error_pct='0.25'))"))
+            .hasMessageContaining("The parameters precision, tree_levels, and distance_error_pct are not applicable to BKD tree indexes.")
+            .hasPGError(INTERNAL_ERROR)
+            .hasHTTPError(BAD_REQUEST, 4000);
+        Asserts.assertSQLError(() -> execute("create table test (col1 geo_shape INDEX using BKDTREE with (tree_levels=2))"))
+            .hasMessageContaining("The parameters precision, tree_levels, and distance_error_pct are not applicable to BKD tree indexes.")
+            .hasPGError(INTERNAL_ERROR)
+            .hasHTTPError(BAD_REQUEST, 4000);
+    }
```
Please check https://github.com/crate/crate/pull/14942#discussion_r1370380600

piotrrzysko replied:
Done. I replied in the thread you mentioned.

BaurzhanSakhariev commented on server/src/test/java/io/crate/lucene/WithinQueryBuilderTest.java:
```diff
@@ -24,18 +24,28 @@
+import org.apache.lucene.document.ShapeField;
+import org.apache.lucene.search.ConstantScoreQuery;
 import org.apache.lucene.search.Query;
 import org.apache.lucene.spatial.prefix.WithinPrefixTreeQuery;
 import org.junit.Test;
 
 public class WithinQueryBuilderTest extends LuceneQueryBuilderTest {
 
```
Was going to say that we need an adjustment in the `WithinFunction.toQuery` but then realized that we do that optimization only for `within(point, geo_shape_literal)`, ie shape is provided by user and is not read from the index so no adjustment is needed. 

Maybe worth adding a code comment next to https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/expression/scalar/geo/WithinFunction.java#L183 that Lucene query construction is index type agnostic for geo shapes since it's provided as Literal

piotrrzysko commented on server/src/main/java/org/elasticsearch/common/geo/builders/GeometryCollectionBuilder.java:
```diff
@@ -163,6 +164,25 @@ public Shape build() {
+                shapes.add(o);
+            }
+        }
+
+        if (shapes.size() == 1) {
+            return shapes.get(0);
+        }
```
I'm aware of that. I wanted to check first why they removed it. I'll address this. Thanks for the reminder :).

piotrrzysko replied:
In my opinion, this check doesn't break anything and may serve as a minor optimization, slightly reducing the number of allocated objects. There is a comment attached to this change: https://github.com/elastic/elasticsearch/pull/35320/files#r240309129. However, shapes collected in the `shapes` list cannot be of the double[] type.

BaurzhanSakhariev replied:
Thanks for investigating!

piotrrzysko commented:
Thanks for the review.

Here's a brief update from me:
* I've addressed some of the review comments.
* I've started working on adding more tests to cover a wider range of shapes and match types. I've discovered that, in some cases, BKD-backed queries return different results than their prefix-backed counterparts. For example:

```sql
-- BKD
create table tbl (id long primary key, shape geo_shape index using bkdtree);
insert into tbl (id, shape) values (1, 'POINT(13.0 52.4)');
select id from tbl where match(shape, 'POINT(13.0 52.4)') using intersects order by id; -- returns 0 rows

-- Prefix
create table tbl_prefix (id long primary key, shape geo_shape);
insert into tbl_prefix (id, shape) values (1, 'POINT(13.0 52.4)');
select id from tbl_prefix where match(shape, 'POINT(13.0 52.4)') using intersects order by id; -- returns 1 row
```

I suspect this is due to precision limitations, but I'll try to investigate this and report back my findings.

piotrrzysko commented:
I've added tests to verify that all available geo match types and shapes work as expected with both prefix and BKD tree-based indexes. Along the way, I discovered two properties of BKD-based indexing implemented in Lucene:

1. WITHIN queries with line geometries are currently not supported. See:
    * [io.crate.integrationtests.BkdTreeGeoMatchIntegrationTest#test_geo_match_within_does_not_support_linestring](https://github.com/crate/crate/pull/14942/files#diff-eba6df8bcdbdf888de3e5ebfc29c23f9776cad5236ac83833c7e25dc3dba140dR22)
    * [LatLonShapeQuery#validateGeometries](https://github.com/apache/lucene/blob/releases/lucene/9.8.0/lucene/core/src/java/org/apache/lucene/document/LatLonShapeQuery.java#L51-L57)
2. The issue related to disparities between results returned by BKD tree and Prefix tree indexes that I mentioned earlier appears to be caused by the fact that Lucene internally encodes dimensions by converting them from 64-bit doubles to 32-bit integers. See: 
    * [GeoEncodingUtils#encodeLatitude](https://github.com/apache/lucene/blob/releases/lucene/9.8.0/lucene/core/src/java/org/apache/lucene/geo/GeoEncodingUtils.java#L51-L65) 
    * https://github.com/apache/lucene/issues/9680). 

   According to the comment in [LatLonPoint](https://github.com/apache/lucene/blob/releases/lucene/9.8.0/lucene/core/src/java/org/apache/lucene/document/LatLonPoint.java#L65-L77) the precision of original shapes is maintained with around 1 cm accuracy. The following tests prove this: 

    ```java
    @Test
    public void fails() {
        execute("create table t (s geo_shape index using bkdtree) with (number_of_replicas=0)");
        ensureGreen();
        execute("insert into t (s) values ('POINT(13.5 52.5)')");
        execute("refresh table t");

        execute("select * from t where match(s, 'POINT(13.5 52.5)')");
        assertThat(response).hasRowCount(1);
    }

    @Test
    public void passes() {
        execute("create table t (s geo_shape index using bkdtree) with (number_of_replicas=0)");
        ensureGreen();
        execute("insert into t (s) values ('POINT(13.5 52.5)')");
        execute("refresh table t");

        double longitude = GeoEncodingUtils.decodeLongitude(GeoEncodingUtils.encodeLongitude(13.5)); // 13.499999949708581
        double latitude = GeoEncodingUtils.decodeLatitude(GeoEncodingUtils.encodeLatitude(52.5)); // 52.49999997206032
        execute("select * from t where match(s, 'POINT(" + longitude + " " + latitude + ")')");
        assertThat(response).hasRowCount(1);
    }
    ```

Please let me know if you would like me to update the documentation by describing what I've already implemented, or if you prefer to explore a different approach to introducing this feature, as mentioned in https://github.com/crate/crate/pull/14942#discussion_r1373028073.

BaurzhanSakhariev commented:
>-- Prefix
create table tbl_prefix (id long primary key, shape geo_shape);
insert into tbl_prefix (id, shape) values (1, 'POINT(13.0 52.4)');
select id from tbl_prefix where match(shape, 'POINT(13.0 52.4)') using intersects order by id; -- returns 1 row

this is BTW might be solution for https://github.com/crate/crate/issues/10604 if it returns 1 row.

With BKD we will have to use distance I guess (as mentioned in the linked ticket)

BaurzhanSakhariev commented:
Hi @piotrrzysko, thanks for addressing comments and adding more tests!

> Please let me know if you would like me to update the documentation by describing what I've already implemented, or if you prefer to explore a different approach to introducing this feature, as mentioned in https://github.com/crate/crate/pull/14942#discussion_r1373028073.

After internal discussion we came to the conclusion that we should integrate BKD indexing as a new indexing strategy but don't make it default. Instead we should keep current default (geohash). 

The reason for keeping old default is to not break some scenarios you mentioned in https://github.com/crate/crate/pull/14942#issuecomment-1793800358. 

On  1: We could fall back to a slower `GenericFunctionQuery` (do full scan) by tweaking `MatchPredicate` to make within/LineString work. But we don't need to do it now because problem 2 is still open (see below).
On 2. No concrete quick workaround for now. Once https://github.com/apache/lucene/issues/9680 is resolved upstream, we can consider making BKD indexing default. 

Regarding deprecation and further removal of the geohash/quadtree in CrateDB 6.0 mentioned in https://github.com/crate/crate/pull/14942#discussion_r1373028073:

Please ignore it, since we keep the current default, we don't deprecate it yet. Also, we of course would allow mixing different indexing strategies in a single table. Please check comment https://github.com/crate/crate/pull/14942#discussion_r1370407987

piotrrzysko commented:
I updated the docs and added info about the change to the release notes.

BaurzhanSakhariev commented on server/src/test/java/io/crate/integrationtests/GeoShapeIntegrationTest.java:
```diff
@@ -58,35 +63,41 @@ public class GeoShapeIntegrationTest extends IntegTestCase {
 
     @Before
     public void prepare() throws Exception {
-        execute("CREATE TABLE shaped (" +
+        for (var table : TABLE_TO_INDEX_TYPE_MAP.entrySet()) {
+            var tableName = table.getKey();
```
Could you please  create all 4 columns (default geohash, geohash array, BKD, BKD array) in a single table?
Assertions are same anyway but we would also test how we can mix different index types in a single table.

piotrrzysko replied:
Sure. Done.

BaurzhanSakhariev reviewed (approved):
Great work, thanks!
